driver/pipeline/middleware.h

```diff
@@ -0,0 +1,259 @@
+// Copyright 2024 Synnax Labs, Inc.
+//
+// Use of this software is governed by the Business Source License included in the file
+// licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with the Business Source
+// License, use of this software will be governed by the Apache License, Version 2.0,
+// included in the file licenses/APL.txt.
+
+#pragma once
+
+#include <memory>
+#include <functional>
+#include <vector>
+#include <map>
+#include <string>
+#include <variant>
+#include <thread>
+
+#include "client/cpp/synnax.h"
+#include "driver/config/config.h"
+
+namespace pipeline {
+    ///////////////////////////////////////////////////////////////////////////////////
+    //                                    Middleware                                 //
+    ///////////////////////////////////////////////////////////////////////////////////
+    class Middleware {
+    public:
+        virtual ~Middleware() = default;
+        virtual bool handle(Frame& frame) = 0;
+    }; // class Middleware
+
+    ///////////////////////////////////////////////////////////////////////////////////
+    //                                  MiddlewareChain                              //
+    ///////////////////////////////////////////////////////////////////////////////////
+    class MiddlewareChain {
+    public:
+        void add(std::shared_ptr <Middleware> middleware) {
+            middlewares.push_back(middleware);
+        }
+
+       bool empty() {
+            return middlewares.empty();
+        }
+
+        freighter::Error exec(Frame &frame) {
+            for (auto &middleware: middlewares) {
+                if (!middleware->handle(frame)) {
+                    return freighter::Error("Middleware failed");
+                }
+            }
+            return freighter::NIL;
+        }
+
+    private:
+        std::vector <std::shared_ptr<Middleware>> middlewares;
+    }; // class MiddlewareChain
+
+    ///////////////////////////////////////////////////////////////////////////////////
+    //                                  TareMiddleware                               //
+    ///////////////////////////////////////////////////////////////////////////////////
+    //TODO: this needs to be the first middleware in the chain (somehow check/force that)?
+    class TareMiddleware : public Middleware {
+    public:
+        explicit TareMiddleware(std::vector<synnax::ChannelKey> keys) {
+            for (auto &key: keys) {
+                tare_values[key] = 0.0;
+            }
+        }
+
+        //TODO: error handling for invalid keys/ keys that weren't configured upon construction
+        void tare(json &arg) {
```

Comment:
I think it would be better to pass in a parser here no?

Reply:
Means we can validate the field here.

Reply:
pass it in here or just construct a parser in here